DynamicTablesPkg/Library/Acpi/Common/AcpiSlitLib/SlitGenerator.c

```diff
@@ -0,0 +1,412 @@
+      if (DomainRelationInfo[Index].Relation != 0x10) {
+        DEBUG ((
+          DEBUG_ERROR,
+          "ERROR: SLIT: Invalid relation value %d for same domain ID %d\n",
+          DomainRelationInfo[Index].Relation,
+          DomainInfoFirst->DomainId
+          ));
+        return EFI_INVALID_PARAMETER;
+      }
+    }
+
+    NumberOfSystemLocalities = MAX (
+                                 NumberOfSystemLocalities,
+                                 MAX (
+                                   DomainInfoFirst->DomainId,
+                                   DomainInfoSecond->DomainId
+                                   )
+                                 );
+  }
+
+  /// Since locality IDs start from 0, increment the count by 1
+  NumberOfSystemLocalities++;
+
+  AcpiSlitTable = AllocateZeroPool (
+                    sizeof (EFI_ACPI_6_5_SYSTEM_LOCALITY_DISTANCE_INFORMATION_TABLE_HEADER) +
+                    (sizeof (UINT8) * NumberOfSystemLocalities * NumberOfSystemLocalities)
+                    );
+  if (AcpiSlitTable == NULL) {
+    DEBUG (
+      (DEBUG_ERROR, "ERROR: SLIT: Failed to allocate memory for SLIT table.\n"));
+    return EFI_OUT_OF_RESOURCES;
+  }
+
+  SlitEntry = (UINT8 *)(AcpiSlitTable + sizeof (EFI_ACPI_6_5_SYSTEM_LOCALITY_DISTANCE_INFORMATION_TABLE_HEADER));
+
+  for (Index = 0; Index < NumberOfSystemLocalities * NumberOfSystemLocalities; Index++) {
+    /// Set the non-diagonal elements to 0xFF
+    SlitEntry[Index] = 0xFF;
+    /// Set the diagonal elements to 0x10
+    if ((Index % (NumberOfSystemLocalities + 1)) == 0) {
+      SlitEntry[Index] = 0x10;
+    }
+  }
+
+  /// Fill the SlitEntry with the Proximity Domain relation information
+  for (Index = 0; Index < DomainRelInfoCount; Index++) {
+    Status = GetEArchCommonObjProxDomainInfo (
+               CfgMgrProtocol,
+               DomainRelationInfo[Index].FirstDomainToken,
+               &DomainInfoFirst,
+               NULL
+               );
+    if (EFI_ERROR (Status)) {
+      DEBUG ((
+        DEBUG_ERROR,
+        "ERROR: SLIT: Failed to retrieve First Proximity Domain information. Status = %r\n",
+        Status
+        ));
+      goto return_error;
+    }
+
+    Status = GetEArchCommonObjProxDomainInfo (
+               CfgMgrProtocol,
+               DomainRelationInfo[Index].SecondDomainToken,
+               &DomainInfoSecond,
+               NULL
+               );
+    if (EFI_ERROR (Status)) {
+      DEBUG ((
+        DEBUG_ERROR,
+        "ERROR: SLIT: Failed to retrieve Second Proximity Domain information. Status = %r\n",
+        Status
+        ));
+      goto return_error;
+    }
+
+    /// fill the SlitEntry with the relation value
+    SlitEntry[DomainInfoFirst->DomainId * NumberOfSystemLocalities + DomainInfoSecond->DomainId] =
+      (UINT8)DomainRelationInfo[Index].Relation;
+
+    /// Verify if the corresponding diagonal element is 0xFF, and if so, populate it with the relation value

```

Comment:
I think that for the 2 indexes in the array, we should check that the values are 0xFF before writing down the new value. Otherwise it means that there are duplicated values provided by the user.